File: Merge k Sorted Lists - Leetcode 23/Merge k Sorted Lists - Leetcode 23.java
class Solution {
    // Time: O(n log k)
    // Space: O(k)
    public ListNode mergeKLists(Listnode[] lists){
        PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a,b)-> Integer(a.val,b.val));

        for(ListNode node:lists){
            if(node!= null){
                minHeap.offer(node);
            }
        }

        ListNode dummy = new ListNode(0)
        ListNode cur = dummy

        while(!minheap.isEmpty()){
            ListNode node = minHeap.poll();
            cur.next = node;
            cur = node;
            if(node.next != null){
                minHeap.offer(node.next);
            }
        }

        return dummy.next;
    }
}

File: Merge k Sorted Lists - Leetcode 23/Merge k Sorted Lists - Leetcode 23.py
class Solution:
    #  Time: O(n log k)
    #  Space: O(k)
    def mergeKlists(self, lists: List[Optional[ListNode]])->Optional[ListNode]:
        heap = []
        for i, node in enumerate(lists):
            heapq.heappush(heap, (node.val, i, node))
        
        D = ListNode()
        cur = D

        while heap:
            val, i, node = heapq.heappop(heap)
            cur.next = node
            cur = node
            if node:
                heapq.heappush(heap, (node.val, i, node))
            
        return D.next

